doc/source/Code-Style-Guide.rst

```diff
@@ -0,0 +1,746 @@
+Exceptions raised. \* Returns: Variables returned by the function or
+method. \* References: Link to available documentation. \* Warns:
+Warnings raised.
+
+“Napolean Google style” Returns: only displays properly in Sphinx if
+there is only one variable being returned. If more than one variable is
+being returned then the return value should be a list of each item
+returned in the tuple. See the example below.
+
+Type annotations
+~~~~~~~~~~~~~~~~
+
+All function and method interfaces, except for those in CLIs (see
+below), should have `type
+annotations <https://docs.python.org/3/library/typing.html>`__. Type
+annotations have been part of Python since version 3.5. Here is a simple
+example showing type annotations for a function that takes a string and
+returns a string.
+
+.. code:: python
+
+   def greeting(name: str) -> str:
+       return 'Hello ' + name
+
+Types are available from the typing module. For example
+
+.. code:: python
+
+   from typing import List
+
+   def first_in_list_of_str(list_of_str: List[str]) -> str:
+       return list_of_str[0]
+
+If an argument or return value can have multiple types, use Union.
+
+.. code:: python
+
+   from typing import List, Union
+
+   def length_of_str_or_list(arg: Union[str, List[str]]) -> int:
+       return len(arg)
+
+Do not put types in the docstring (except for CLIs, see next section).
+
+.. code:: python
+
+   def greeting(name: str) -> str:
+       """My greeting.
+
+       Args:
+           name:
+               The name to greet.
+
+       Returns:
+           The greeting with appropriate name.
+       """
+       return 'Hello ' + name
+
+If a function or method has multiple returns then the return type is a
+Tuple. The return in the docstring should be a list (in rst/markdown
+style).
+
+.. code:: python
+
+   from typing import Tuple
+
+   def first_and_last(list_of_str: List[str]) -> Tuple[str, str]:
+       """First and last items.
+
+       Args:
+           list_of_str:
+               A list of strings.
+
+       Returns:
+           - First item.
+           - Second item.
+       """
+       return list_of_str[0], list_of_str[-1]
+
+See the plugin example below, and throughout the existing codebase for
+more examples.
+
+Variable types in CLIs
+~~~~~~~~~~~~~~~~~~~~~~
+
+Note that we use clize in the CLIs which uses type annotations at
+runtime, hence the need to define these clearly within the docstring.
+
+Within docstrings, when specifying a variable type, Python built-in data
+types can be used directly e.g. \* int \* float \* str \* bool
+
+In order for variable types to link correctly within
+`readthedocs <http://improver.readthedocs.io/en/latest/?badge=latest>`__,
+the ``intersphinx_mapping`` needs to be updated to link to the
+documentation of the module where the variable type originates. For
+example: \* numpy.ndarray \* datatime.datetime The full name of the
+module is required, rather than an alias e.g. numpy rather than np.
```

Comment:
Another bulleted list that has been inlined.

Reply:
Fixed.